File: core/management/commands/expire_ads.py
"""Expire active advertisements whose paid validity has ended."""
from django.core.management.base import BaseCommand
from django.utils import timezone

from ads.models import Ad, AdStatus
from ads.services.lifecycle import transition


class Command(BaseCommand):
    help = 'Mark active advertisements as expired once expires_at has passed.'

    def add_arguments(self, parser):
        parser.add_argument('--limit', type=int, default=500, help='Maximum advertisements to process.')

    def handle(self, *args, **options):
        now = timezone.now()
        candidates = (
            Ad.objects.filter(
                status=AdStatus.ACTIVE,
                deleted_at__isnull=True,
                expires_at__isnull=False,
                expires_at__lte=now,
            )
            .order_by('expires_at')[:max(options['limit'], 1)]
        )
        count = 0
        for ad in candidates:
            transition(ad, AdStatus.EXPIRED, reason='انقضای خودکار اعتبار آگهی')
            count += 1
        self.stdout.write(self.style.SUCCESS(f'{count} آگهی منقضی شد.'))

File: app/Console/Commands/ExpireAds.php
<?php

declare(strict_types=1);
namespace App\Console\Commands;
use App\Domains\Ads\Enums\AdStatus;
use App\Domains\Ads\Services\AdWorkflow;
use App\Models\Ad;
use Illuminate\Console\Command;
class ExpireAds extends Command { protected $signature='ads:expire {--dry-run : فقط تعداد قابل انقضا را نشان بده}'; protected $description='آگهی‌های فعال منقضی‌شده را با روش idempotent منقضی می‌کند.'; public function handle(AdWorkflow $workflow): int { $query=Ad::query()->where('status',AdStatus::Active)->whereNotNull('expires_at')->where('expires_at','<=',now()); if ($this->option('dry-run')) { $this->info((string)$query->count()); return self::SUCCESS; } $count=0; $query->orderBy('id')->chunkById(100,function($ads) use(&$count,$workflow):void { foreach($ads as $ad){ $workflow->transition($ad,AdStatus::Expired,null,'انقضای زمان‌بندی‌شده'); $count++; } }); $this->info("{$count} آگهی منقضی شد."); return self::SUCCESS; } }
